pkg/prometheus/statefulset.go

```diff
@@ -1080,3 +1097,48 @@ func queryLogFilePath(p *monitoringv1.Prometheus) string {
 
 	return filepath.Join(defaultQueryLogDirectory, p.Spec.QueryLogFile)
 }
+
+func intersection(a, b []string) (i []string) {
+	m := make(map[string]bool)
+
+	for _, item := range a {
+		m[item] = true
+	}
+
+	for _, item := range b {
+		if _, ok := m[item]; ok {
+			i = append(i, item)
+		}
+	}
+	return i
+}
+
+func extractArgKeys(args []string) ([]string, []string) {
+	var k []string
+	var e []string
+	for _, arg := range args {
+		key := argKeyRegex.FindStringSubmatch(arg)
+		if len(key) != 2 {
+			e = append(e, arg)
+		} else {
+			k = append(k, key[1])
+		}
+	}
+
+	return k, e
+}
+
+func verifyAdditionalArgs(operatorArgs []string, additionalArgs []string) (e error) {
+	operatorArgKeys, _ := extractArgKeys(operatorArgs)
+	additionalArgKeys, invalidArgs := extractArgKeys(additionalArgs)
+	if len(invalidArgs) > 0 {
+		return errors.Errorf("invalid argument syntax for args: %s", invalidArgs)
+	}
+
+	i := intersection(operatorArgKeys, additionalArgKeys)
+	if len(i) > 0 {
+		return errors.Errorf("invalid additionalArgs configuration for already defined args: %s", i)
```

Comment:
There are 2 options
* Fail the reconciliation in case of conflict (as you did here).
* Silently drop the additional argument which is already managed by the operator.

I agree that an explicit failure is better (users won't have to wonder why the operator dropped the additional arguments) but it also means that when a new version of the operator starts managing an argument, the reconciliation will break right after the upgrade. At least, the behavior should be documented in the API.

Reply:
Extended docs